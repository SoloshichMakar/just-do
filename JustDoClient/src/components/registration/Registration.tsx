import React from "react";
import { gql } from "apollo-boost";
import { useMutation } from "@apollo/react-hooks";
import "../style/style.scss";
import {
  IconButton,
  Snackbar,
} from "@material-ui/core";
import JustDoComponent from "../../picture/JustDoComponent.png";
import JustDoLogo from "../../picture/JustDoLogo.png";
import {
  emailHelperMessage,
  passwordHelperMessage,
  confirmPasswordHelperMessage,
  validateEmail,
  validatePassword,
  validateConfirmPassword,
} from "../../utils/validateUserDataFunction";
import Visibility from "@material-ui/icons/Visibility";
import VisibilityOff from "@material-ui/icons/VisibilityOff";
import Input from "@material-ui/core/Input";
import InputAdornment from "@material-ui/core/InputAdornment";
import FormControl from "@material-ui/core/FormControl";
import { Alert } from "@material-ui/lab";
import {NavLink, useHistory} from "react-router-dom";
import { ColorButton, CustomLink } from "../style/CustomElements";


interface IRegistrationProps {
  email: string;
  password: string;
  message: string;
  showPassword: boolean;
  showConfirmPassword: boolean;
  confirmPassword: string;
  setMessage: Function;
  formClear: Function;
  errorMessage: Function;
  createUserMessage: Function;
  clearMessage: Function;
  textEmailChange: Function;
  textPasswordChange: Function;
  textConfirmPasswordChange: Function;
  showPasswordValue: Function;
  showConfirmPasswordValue: Function;
}

const ADD_USER = gql`
  mutation AddUser($email: String!, $password: String!) {
    addUser(email: $email, password: $password) {
      id
      email
    }
  }
`;

let emailIsEdit = false;
let passwordIsEdit = false;
let confirmPasswordIsEdit = false;

const Registration: React.FC<IRegistrationProps> = ({
  email,
  password,
  message,
  confirmPassword,
  showPassword,
  clearMessage,
  errorMessage,
  formClear,
  setMessage,
  createUserMessage,
  showConfirmPassword,
  textEmailChange,
  textPasswordChange,
  showPasswordValue,
  showConfirmPasswordValue,
  textConfirmPasswordChange,
}) => {
  const [addUser] = useMutation(ADD_USER);
  const history = useHistory();

  function currentTextEmailChange(e: React.ChangeEvent<HTMLInputElement>) {
    emailIsEdit = true;
    const getTextAreaValue = e.target.value;
    textEmailChange(getTextAreaValue);
  }

  function redirectToSighIn() {
    formClear();
    emailIsEdit = false;
    passwordIsEdit = false;
  }

  const handleClickShowPassword = () => {
    showPasswordValue(showPassword);
  };

  const handleClickShowConfirmPassword = () => {
    showConfirmPasswordValue(showConfirmPassword);
  };

  function currentTextPasswordChange(e: React.ChangeEvent<HTMLInputElement>) {
    passwordIsEdit = true;
    const getTextAreaValue = e.target.value;
    textPasswordChange(getTextAreaValue);
  }

  function currentTextConfirmPasswordChange(
    e: React.ChangeEvent<HTMLInputElement>
  ) {
    confirmPasswordIsEdit = true;
    const getTextAreaValue = e.target.value;
    textConfirmPasswordChange(getTextAreaValue);
  }

  async function createNewUser() {
    try {
      if (
        !validateEmail(emailIsEdit, email) &&
        emailIsEdit &&
        !validatePassword(passwordIsEdit, password) &&
        passwordIsEdit &&
        !validateConfirmPassword(
          confirmPasswordIsEdit,
          password,
          confirmPassword
        ) &&
        confirmPasswordIsEdit
      ) {
        await addUser({
          variables: { email: email, password: password },
        });
        emailIsEdit = false;
        passwordIsEdit = false;
        confirmPasswordIsEdit = false;
        setMessage("User: " + email + " is created");
        createUserMessage();
        history.push("/login");
      }
    } catch (error) {
      errorMessage();
    }
  }

  const handleErrorClose = (event?: React.SyntheticEvent, reason?: string) => {
    if (reason === "clickaway") {
      return;
    }
    clearMessage();
  };

  return (
    <div id="registration_main">
      <Snackbar
        open={message.length !== 0}
        autoHideDuration={10000}
        onClose={handleErrorClose}
      >
        <Alert onClose={handleErrorClose} severity="error">
          {message}
        </Alert>
      </Snackbar>
      <div id="registration_main__logo_images">
        <img id="logo" src={JustDoLogo} />
        <img src={JustDoComponent} id="background_image" />
      </div>
      <div id="registration_main__components">
        <h1>Sign up</h1>
        <div id="components__input_components">
          <div className="input_components__text_input">
            <FormControl id="textInput__email">
              <Input
                id="email_input"
                error={validateEmail(emailIsEdit, email)}
                type="text"
                value={email}
                onChange={currentTextEmailChange}
                placeholder="E-mail"
              />
              <div id="errorMessage">
                {emailHelperMessage(emailIsEdit, email)}
              </div>
            </FormControl>
          </div>

          <div className="input_components__text_input">
            <FormControl id="textInput__passwordForm">
              <Input
                id="password_input"
                error={validatePassword(passwordIsEdit, password)}
                type={showPassword ? "text" : "password"}
                value={password}
                placeholder="Password"
                onChange={currentTextPasswordChange}
                endAdornment={
                  <InputAdornment position="end">
                    <IconButton
                      aria-label="toggle password visibility"
                      onClick={handleClickShowPassword}
                    >
                      {showPassword ? <Visibility /> : <VisibilityOff />}
                    </IconButton>
                  </InputAdornment>
                }
              />
              <div id="errorMessage">
                {passwordHelperMessage(passwordIsEdit, password)}
              </div>
            </FormControl>
          </div>

          <div className="input_components__text_input">
            <FormControl id="textInput__passwordConfirmForm">
              <Input
                id="password_confirm_input"
                error={validateConfirmPassword(
                  passwordIsEdit,
                  password,
                  confirmPassword
                )}
                type={showConfirmPassword ? "text" : "password"}
                value={confirmPassword}
                onChange={currentTextConfirmPasswordChange}
                placeholder="Confirm password"
                endAdornment={
                  <InputAdornment position="end">
                    <IconButton
                      aria-label="toggle password visibility"
                      onClick={handleClickShowConfirmPassword}
                    >
                      {showConfirmPassword ? <Visibility /> : <VisibilityOff />}
                    </IconButton>
                  </InputAdornment>
                }
              />
              <div id="errorMessage">
                {confirmPasswordHelperMessage(
                  confirmPasswordIsEdit,
                  password,
                  confirmPassword
                )}
              </div>
            </FormControl>
          </div>
        </div>

        <div id="input_components__color_button">
          <ColorButton
            id="color_button"
            variant="contained"
            color="primary"
            onClick={() => createNewUser()}
          >
            Sign up
          </ColorButton>
        </div>
        <div id="input_components__footer__sign_in">
          I already have an account.
          <NavLink to="/login"><CustomLink onClick={() => redirectToSighIn()}>Sign In</CustomLink></NavLink>
        </div>
        <div id="input_components__footer__agreements">
          <div>By accessing your account, you agree to our</div>
          <div id="agreements__links">
            <NavLink to="/#"><CustomLink
              id="item__terms_conditions"
              className="agreements__links__item"
            >
              Terms conditions
            </CustomLink></NavLink>
            <div className="agreements__links__item"> and </div>
            <NavLink to="/#"><CustomLink
              id="item__privacy_policy"
              className="agreements__links__item"
            >
              Privacy Policy
            </CustomLink></NavLink>
          </div>
        </div>
      </div>
    </div>
  );
};

export default Registration;
