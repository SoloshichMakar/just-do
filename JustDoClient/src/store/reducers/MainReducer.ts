import RegistrationReducer from "./RegistrationReducer";
import AuthorizationReducer from "./AuthorizationReducer";
import PasswordRestoreReducer from "./PasswordRestoreReducer";
import ForgotPasswordReducer from "./ForgotPasswordReducer";

import { combineReducers } from "redux";

const MainReducer = combineReducers({
  RegistrationReducer,
  AuthorizationReducer,
  PasswordRestoreReducer,
  ForgotPasswordReducer,
});

export default MainReducer;
