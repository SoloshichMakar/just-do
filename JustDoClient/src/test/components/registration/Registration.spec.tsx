import renderer from "react-test-renderer";
import { createStore } from "redux";
import MainReducer from "../../../store/reducers/MainReducer";
import { Provider } from "react-redux";
import React from "react";
import Registration from "../../../components/registration/Registration";
import ApolloClient from "apollo-boost";
import { ApolloProvider } from "react-apollo";
import {BrowserRouter} from "react-router-dom";

const props = {
  email: "",
  password: "",
  message: "",
  showPassword: false,
  showConfirmPassword: false,
  confirmPassword: "",
  setMessage: Function,
  errorMessage: Function,
  formClear: Function,
  createUserMessage: Function,
  clearMessage: Function,
  textEmailChange: Function,
  textPasswordChange: Function,
  textConfirmPasswordChange: Function,
  showPasswordValue: Function,
  showConfirmPasswordValue: Function,
};

const client = new ApolloClient({
  uri: ``,
});
const store = createStore(MainReducer);

describe("Registration snapshot", () => {
  it("Registration render correctly", () => {
    const tree = renderer
      .create(
        <ApolloProvider client={client}>
          <Provider store={store}>
            <BrowserRouter>
            <Registration
              email={props.email}
              password={props.password}
              message={props.message}
              showPassword={props.showPassword}
              showConfirmPassword={props.showConfirmPassword}
              confirmPassword={props.confirmPassword}
              errorMessage={props.errorMessage}
              createUserMessage={props.createUserMessage}
              clearMessage={props.clearMessage}
              textEmailChange={props.textEmailChange}
              textPasswordChange={props.textPasswordChange}
              textConfirmPasswordChange={props.textConfirmPasswordChange}
              showPasswordValue={props.showPasswordValue}
              showConfirmPasswordValue={props.showConfirmPasswordValue}
              formClear={props.formClear}
              setMessage={props.setMessage}
            />
            </BrowserRouter>
          </Provider>
        </ApolloProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
