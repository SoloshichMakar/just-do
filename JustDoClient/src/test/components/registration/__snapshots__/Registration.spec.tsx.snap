// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Registration snapshot Registration render correctly 1`] = `
<div
  id="registration_main"
>
  <div
    id="registration_main__logo_images"
  >
    <img
      id="logo"
      src="JustDoLogo.png"
    />
    <img
      id="background_image"
      src="JustDoComponent.png"
    />
  </div>
  <div
    id="registration_main__components"
  >
    <h1>
      Sign up
    </h1>
    <div
      id="components__input_components"
    >
      <div
        className="input_components__text_input"
      >
        <div
          className="MuiFormControl-root"
          id="textInput__email"
        >
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              className="MuiInputBase-input MuiInput-input"
              disabled={false}
              id="email_input"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="E-mail"
              required={false}
              type="text"
              value=""
            />
          </div>
          <div
            id="errorMessage"
          />
        </div>
      </div>
      <div
        className="input_components__text_input"
      >
        <div
          className="MuiFormControl-root"
          id="textInput__passwordForm"
        >
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              className="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
              disabled={false}
              id="password_input"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Password"
              required={false}
              type="password"
              value=""
            />
            <div
              className="MuiInputAdornment-root MuiInputAdornment-positionEnd"
            >
              <button
                aria-label="toggle password visibility"
                className="MuiButtonBase-root MuiIconButton-root"
                disabled={false}
                id="passwordVisibility"
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 7c2.76 0 5 2.24 5 5 0 .65-.13 1.26-.36 1.83l2.92 2.92c1.51-1.26 2.7-2.89 3.43-4.75-1.73-4.39-6-7.5-11-7.5-1.4 0-2.74.25-3.98.7l2.16 2.16C10.74 7.13 11.35 7 12 7zM2 4.27l2.28 2.28.46.46C3.08 8.3 1.78 10.02 1 12c1.73 4.39 6 7.5 11 7.5 1.55 0 3.03-.3 4.38-.84l.42.42L19.73 22 21 20.73 3.27 3 2 4.27zM7.53 9.8l1.55 1.55c-.05.21-.08.43-.08.65 0 1.66 1.34 3 3 3 .22 0 .44-.03.65-.08l1.55 1.55c-.67.33-1.41.53-2.2.53-2.76 0-5-2.24-5-5 0-.79.2-1.53.53-2.2zm4.31-.78l3.15 3.15.02-.16c0-1.66-1.34-3-3-3l-.17.01z"
                    />
                  </svg>
                </span>
              </button>
            </div>
          </div>
          <div
            id="errorMessage"
          />
        </div>
      </div>
      <div
        className="input_components__text_input"
      >
        <div
          className="MuiFormControl-root"
          id="textInput__passwordConfirmForm"
        >
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              className="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
              disabled={false}
              id="password_confirm_input"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Confirm password"
              required={false}
              type="password"
              value=""
            />
            <div
              className="MuiInputAdornment-root MuiInputAdornment-positionEnd"
            >
              <button
                aria-label="toggle password visibility"
                className="MuiButtonBase-root MuiIconButton-root"
                disabled={false}
                id="confirmPasswordVisibility"
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 7c2.76 0 5 2.24 5 5 0 .65-.13 1.26-.36 1.83l2.92 2.92c1.51-1.26 2.7-2.89 3.43-4.75-1.73-4.39-6-7.5-11-7.5-1.4 0-2.74.25-3.98.7l2.16 2.16C10.74 7.13 11.35 7 12 7zM2 4.27l2.28 2.28.46.46C3.08 8.3 1.78 10.02 1 12c1.73 4.39 6 7.5 11 7.5 1.55 0 3.03-.3 4.38-.84l.42.42L19.73 22 21 20.73 3.27 3 2 4.27zM7.53 9.8l1.55 1.55c-.05.21-.08.43-.08.65 0 1.66 1.34 3 3 3 .22 0 .44-.03.65-.08l1.55 1.55c-.67.33-1.41.53-2.2.53-2.76 0-5-2.24-5-5 0-.79.2-1.53.53-2.2zm4.31-.78l3.15 3.15.02-.16c0-1.66-1.34-3-3-3l-.17.01z"
                    />
                  </svg>
                </span>
              </button>
            </div>
          </div>
          <div
            id="errorMessage"
          />
        </div>
      </div>
    </div>
    <div
      id="input_components__color_button"
    >
      <button
        className="MuiButtonBase-root MuiButton-root WithStyles(ForwardRef(Button))-root-72 MuiButton-contained MuiButton-containedPrimary"
        disabled={false}
        id="color_button"
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Sign up
        </span>
      </button>
    </div>
    <div
      id="input_components__footer__sign_in"
    >
      I already have an account.
      <a
        className="MuiTypography-root MuiLink-root WithStyles(ForwardRef(Link))-root-102 MuiLink-underlineHover MuiTypography-colorPrimary"
        href="/login"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
      >
        Sign In
      </a>
    </div>
    <div
      id="input_components__footer__agreements"
    >
      <div>
        By accessing your account, you agree to our
      </div>
      <div
        id="agreements__links"
      >
        <a
          className="MuiTypography-root MuiLink-root WithStyles(ForwardRef(Link))-root-102 MuiLink-underlineHover agreements__links__item MuiTypography-colorPrimary"
          href="/#"
          id="item__terms_conditions"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          Terms conditions
        </a>
        <div
          className="agreements__links__item"
        >
           and 
        </div>
        <a
          className="MuiTypography-root MuiLink-root WithStyles(ForwardRef(Link))-root-102 MuiLink-underlineHover agreements__links__item MuiTypography-colorPrimary"
          href="/#"
          id="item__privacy_policy"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          Privacy Policy
        </a>
      </div>
    </div>
  </div>
</div>
`;
